### PR TITLE
New Function of Create Android Assignment(resolves #135)

### LINKED ISSUE #135: Add Android Espresso into ProgEdu System
1. The teacher can create an android assignment.

### PR DESCRIPTION
**Pull request by MomentaryChen**
1. Add an android emulator in Jenkins automatically.
2. New UI Failure Status.
3. Testing android assignment(done).

### CODE REVIEW

File: server/front/src/app/views/shared/assignment-choose/assignment-choose.component.ts [hunk 1]
```diff
@@ -94,7 +93,9 @@ export class AssignmentChoosedComponent implements OnInit {
       response => {
         this.feedbacks = response;
         this.selectedCommitNumber = commitNumber;
-        this.getScreenshotUrls();
+        if (this.assignment.type === 'WEB' || this.assignment.type === 'ANDROID') {
```
Comment (hoky841023) [inline server/front/src/app/views/shared/assignment-choose/assignment-choose.component.ts]: isShowScreenshot()

File: server/src/main/java/fcu/selab/progedu/project/AndroidAssignment.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+      Transformer transformer = transformerFactory.newTransformer();
+      DOMSource source = new DOMSource(doc);
+      StreamResult result = new StreamResult(new File(jenkinsJobConfigPath));
+      transformer.transform(source, result);
+    } catch (LoadConfigFailureException
+        | ParserConfigurationException
+        | SAXException
+        | IOException
+        | TransformerException e) {
+      e.printStackTrace();
```
Comment (hoky841023) [inline server/src/main/java/fcu/selab/progedu/project/AndroidAssignment.java]: log

File: server/src/main/java/fcu/selab/progedu/project/AndroidAssignment.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    }
+  }
+
+  @Override
+  public void createTemplate(String uploadDirectory) {
+    try {
+      FileUtils.deleteDirectory(new File(uploadDirectory + "/app/src/androidTest/java"));
```
Comment (hoky841023) [inline server/src/main/java/fcu/selab/progedu/project/AndroidAssignment.java:119]: 註解

File: server/src/main/java/fcu/selab/progedu/service/AssignmentService.java [hunk 4]
```diff
@@ -141,22 +141,26 @@ public Response createAssignment(@FormDataParam("assignmentName") String assignm
     // 4. Unzip the uploaded file to tests folder and uploads folder on tomcat,
     // extract main method from tests folder, then zip as root project
     String testDirectory = testDir + assignmentName;
-
+    System.out.println("TEST1");
```
Comment (hoky841023) [inline server/src/main/java/fcu/selab/progedu/service/AssignmentService.java]: ?

File: server/src/main/java/fcu/selab/progedu/service/StatusService.java [hunk 5]
```diff
@@ -122,4 +122,57 @@ public boolean isWebEslintFailure(String console) {
...
   }
+
+  /**
+   * Android is compile error
+   *
+   * @param console jenkins job console text
+   * @return boolean
+   */
+  public boolean isAndroidCompileFailure(String console) {
+    if (console.contains("Task :app:compileDebugJavaWithJavac FAILED")) {
```
Comment (hoky841023) [inline server/src/main/java/fcu/selab/progedu/service/StatusService.java]: return console.contains()

File: server/src/main/java/fcu/selab/progedu/status/AndroidCheckstyleFailure.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  public String extractFailureMsg(String consoleText) {
+    String feedback;
+    String feedbackStart = "Task :app:checkStyle";
+    String feedbackEnd = "FAILURE: Build failed with an exception.";
+    feedback = consoleText.substring(consoleText.indexOf(feedbackStart) + feedbackStart.length(),
+            consoleText.indexOf(feedbackEnd));
+    /**
+     * Remove /var/jenkins_home/workspace/
+     */
+    feedback = feedback.replaceAll("/var/jenkins_home/workspace", "").trim();
```
Comment (hoky841023) [inline server/src/main/java/fcu/selab/progedu/status/AndroidCheckstyleFailure.java]: return feedback.replaceAll().trim()

File: server/front/src/app/views/shared/assignment-choose/assignment-choose.component.ts (hunk 1 above)
Comment (MomentaryChen) [reply server/front/src/app/views/shared/assignment-choose/assignment-choose.component.ts]: Thanks, Fixed it already.

File: server/src/main/java/fcu/selab/progedu/status/AndroidCheckstyleFailure.java (hunk 6 above)
Comment (MomentaryChen) [reply server/src/main/java/fcu/selab/progedu/status/AndroidCheckstyleFailure.java]: Thanks.

File: server/src/main/java/fcu/selab/progedu/service/StatusService.java (hunk 5 above)
Comment (MomentaryChen) [reply server/src/main/java/fcu/selab/progedu/service/StatusService.java]: Thanks.

File: server/src/main/java/fcu/selab/progedu/service/AssignmentService.java (hunk 4 above)
Comment (MomentaryChen) [reply server/src/main/java/fcu/selab/progedu/service/AssignmentService.java]: Thanks.

File: server/src/main/java/fcu/selab/progedu/project/AndroidAssignment.java (hunk 3 above)
Comment (MomentaryChen) [reply server/src/main/java/fcu/selab/progedu/project/AndroidAssignment.java:119]: Thanks.

File: server/src/main/java/fcu/selab/progedu/project/AndroidAssignment.java (hunk 2 above)
Comment (MomentaryChen) [reply server/src/main/java/fcu/selab/progedu/project/AndroidAssignment.java]: Thanks.

**Pull request merged**